PlatformIO ESP32 code/OSSM_ESP32/src/Utilities.cpp

```diff
@@ -88,55 +97,119 @@ void OSSM::runStrokeEngine()
     Stroker.begin(&strokingMachine, &servoMotor);
     Stroker.thisIsHome();
 
-    float lastSpeedPercentage = 0;
-    float lastStrokePercentage = 0;
-    bool lastEncoderButtonToggle = encoderButtonToggle;
-    int strokePattern = 0;
+    float lastSpeedPercentage = speedPercentage;
+    float lastStrokePercentage = strokePercentage;
+    float lastDepthPercentage = depthPercentage;
+    float lastSensationPercentage = sensationPercentage;
+    int lastEncoderButtonPresses = encoderButtonPresses;
+    strokePattern = 0;
+    strokePatternCount = Stroker.getNumberOfPattern();
 
-    Stroker.setSensation(-20, true);
+    Stroker.setSensation(calculateSensation(sensationPercentage), true);
 
     Stroker.setPattern(strokePattern, true);
-    Stroker.setDepth(abs(maxStrokeLengthMm), true);
-    Stroker.setStroke(10, true);
-    Stroker.startPattern();
+    Stroker.setDepth(0.01 * depthPercentage * abs(maxStrokeLengthMm), true);
+    Stroker.setStroke(0.01 * strokePercentage * abs(maxStrokeLengthMm), true);
     Serial.println(Stroker.getState());
     g_ui.UpdateMessage(Stroker.getPatternName(strokePattern));
 
     for (;;)
     {
         Serial.println("looping");
-        if (abs(speedPercentage - lastSpeedPercentage) > 5)
+        if (isChangeSignificant(lastSpeedPercentage, speedPercentage))
         {
-            Serial.println("changing speed");
+            Serial.printf("changing speed: %f\n", speedPercentage * 3);
+            if(speedPercentage == 0) {
+                Stroker.stopMotion();
```

Comment:
This is necessary to reliably stop the motion of StrokeEngine. Otherwise the minimum speed appears to be 0.5